### PR TITLE
OPENEUROPA-9

### PR DESCRIPTION
**Pull request by pfrenssen**

### CODE REVIEW

File: grumphp.yml.dist [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  tasks.phpcs.triggered_by:
+    - php
+    - module
+    - inc
+    - theme
+    - install
+
+  # At the moment the Code Review component does not supports adding tasks
+  # so we need to replicate the whole section here.
+  # @see https://webgate.ec.europa.eu/CITnet/jira/browse/OPENEUROPA-550
```
Comment (ademarco) [inline grumphp.yml.dist]: We can now use the `extra_tasks:` available in coder-review ~0.3 so that we don't have to replicate `phpcs`, `phpmd`, `git_commit_message`.

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+version: '2'
+services:
+
+  web:
+    image: fpfis/php71-dev
```
Comment (ademarco) [inline docker-compose.yml]: We should use new Docker images from devops both here and in `.drone.yml`. Check [this PR](https://github.com/openeuropa/oe_theme/pull/101/files) for more info.

File: .drone.yml [hunk 3]
```diff
@@ -0,0 +1,47 @@
+workspace:
+  base: /test
+  path: openeuropa-site
+
+services:
+  web:
+    image: fpfis/php71-dev
```
Comment (ademarco) [inline .drone.yml]: We should use new Docker images here, check [this PR](https://github.com/openeuropa/oe_theme/pull/101/files) for more.

File: behat.yml.dist [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        - Drupal\DrupalExtension\Context\MinkContext
+        - OpenEuropa\my_site\Context\DrupalContext
+        - OpenEuropa\my_site\Context\FeatureContext
+  extensions:
+    Behat\MinkExtension:
+      goutte: ~
+      browser_name: 'chrome'
+      javascript_session: 'selenium2'
+      selenium2:
+        wd_host: 'http://selenium:4444/wd/hub'
```
Comment (ademarco) [inline behat.yml.dist]: On other components we made this configurable, we should do this here too. Check the following:

- [Runner configuration](https://github.com/openeuropa/oe_theme/blob/master/runner.yml.dist#L31)
- [Behat configuration](https://github.com/openeuropa/oe_theme/blob/master/behat.yml.dist#L34)

File: composer.json [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        "openeuropa/oe_profile": "~0.1",
+        "openeuropa/oe_theme": "~0.4",
+        "webflo/drupal-finder": "^1.0.0",
+        "webmozart/path-util": "^2.3"
+    },
+    "require-dev": {
+        "drupal/console": "^1.0.2",
+        "drupal/drupal-extension": "~4.0",
+        "nikic/php-parser": "~2.1|~3.0",
+        "openeuropa/code-review": "~0.2",
```
Comment (ademarco) [inline composer.json]: Once we use `extra_tasks:` in `grumphp.yml.dist` we should change this into `~0.3`

File: composer.json [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            "build/web/modules/contrib/{$name}": ["type:drupal-module"],
+            "build/web/profiles/contrib/{$name}": ["type:drupal-profile"],
+            "build/web/themes/contrib/{$name}": ["type:drupal-theme"]
+        },
+        "patches": {
+            "openeuropa/oe_theme": {
+                "OPENEUROPA-722: Make sure the default body classes are output - https://github.com/openeuropa/oe_theme/pull/108": "https://patch-diff.githubusercontent.com/raw/openeuropa/oe_theme/pull/108.patch"
+            },
+            "phpro/grumphp": {
+                "Make detection of strict types compatible with PhpParser 4 - https://github.com/phpro/grumphp/pull/525": "https://patch-diff.githubusercontent.com/raw/phpro/grumphp/pull/525.patch"
```
Comment (ademarco) [inline composer.json]: Since we are set on `"nikic/php-parser": "~2.1|~3.0"` we don't need this patch here right? Also, the PR above has been merged but not yet released, instead of including this patch here I would open a follow-up issue about upgrading PhpParser to ~4 on all our components as soon as GrumPHP has a new release.

File: composer.json [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        "installer-paths": {
+            "build/web/core": ["type:drupal-core"],
+            "build/web/libraries/{$name}": ["type:drupal-library"],
+            "build/web/modules/contrib/{$name}": ["type:drupal-module"],
+            "build/web/profiles/contrib/{$name}": ["type:drupal-profile"],
+            "build/web/themes/contrib/{$name}": ["type:drupal-theme"]
+        },
+        "patches": {
+            "openeuropa/oe_theme": {
+                "OPENEUROPA-722: Make sure the default body classes are output - https://github.com/openeuropa/oe_theme/pull/108": "https://patch-diff.githubusercontent.com/raw/openeuropa/oe_theme/pull/108.patch"
```
Comment (ademarco) [inline composer.json]: This has been merged but not released, I'll release a new version of the theme so we can remove this patch from here.

File: composer.json [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                }
+            }
+        },
+        "composer-exit-on-patch-failure": true,
+        "installer-paths": {
+            "build/web/core": ["type:drupal-core"],
+            "build/web/libraries/{$name}": ["type:drupal-library"],
+            "build/web/modules/contrib/{$name}": ["type:drupal-module"],
+            "build/web/profiles/contrib/{$name}": ["type:drupal-profile"],
+            "build/web/themes/contrib/{$name}": ["type:drupal-theme"]
```
Comment (ademarco) [inline composer.json]: On all our components we use `./build` as docroot, while I'm not saying that was necessarily a good idea I'd still like to keep consistency between our components and the sites, shall we use `./build` here too?

File: runner.yml.dist [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    user: "root"
+    password: ""
+  site:
+    profile: "oe_profile"
+    name: "My OpenEuropa site"
+  pre_install:
+    - { task: "chmod", permissions: 0775, file: "${drupal.root}/sites/default" }
+    - { task: "touch", file: "${drupal.root}/sites/default/settings.php" }
+    - { task: "chmod", permissions: 0664, file: "${drupal.root}/sites/default/settings.php" }
+    - { task: "remove", file: "${drupal.root}/sites/default/settings.php" }
```
Comment (ademarco) [inline runner.yml.dist]: I'm wondering if these steps shouldn't be ran anyway by the Task Runner, we can always give the option to the user to opt-out of it by setting a parameter or something.

If we adopt this kind of approach on a post install too we could make sure that permissions are properly set after site installation which, incidentally, will fix the docker container permissions issues too.

What do you think? Shall we open a follow up for this?

File: scripts/composer/SetupWizard.php [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    $config['scripts']['post-create-project-cmd'] = array_diff($config['scripts']['post-create-project-cmd'], ['@cleanup']);
+    if (empty($config['scripts']['post-create-project-cmd'])) {
+      unset($config['scripts']['post-create-project-cmd']);
+    }
+    $config['scripts']['post-root-package-install'] = array_diff($config['scripts']['post-root-package-install'], ['@setup']);
+    if (empty($config['scripts']['post-root-package-install'])) {
+      unset($config['scripts']['post-root-package-install']);
+    }
+
+    $composer_json->write($config);
```
Comment (ademarco) [inline scripts/composer/SetupWizard.php]: This is awesome! :+1:

File: scripts/composer/SetupWizard.php [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+    return TRUE;
```
Comment (ademarco) [inline scripts/composer/SetupWizard.php]: Can we run a `composer dump-autoload` here so that class map is regenerated with new name-spacing? I've looked for ways to do programmatically via the Composer API but we have found any.

File: README.md [hunk 12]
```diff
@@ -1,2 +1,126 @@
...
+This is based on [drupal-composer/drupal-project](https://github.com/drupal-composer/drupal-project).
+
+[![Build Status](https://drone.fpfis.eu/api/badges/openeuropa/drupal-site-template/status.svg?branch=master)](https://drone.fpfis.eu/openeuropa/drupal-site-template)
+
+This project template provides a starter kit for creating a website using the
+OpenEuropa Drupal 8 platform. It will install the [OpenEuropa Profile](https://github.com/openeuropa/oe_profile)
+which is a lightweight installation profile that includes the minimal number
+of modules that are required to enable the [OpenEuropa Theme](https://github.com/openeuropa/oe_theme).
+Using this theme will ensure that the project complies with the guidelines for
+European Commission websites.
```
Comment (ademarco) [inline README.md]: The platform will power also non European Commission site, I'd rephrase this with something like:

> ...will ensure that the project complies with the [Europa Component Library](https://github.com/ec-europa/europa-component-library).

File: README.md [hunk 13]
```diff
@@ -1,2 +1,126 @@
...
+
+In order to build the functionality of the website you are free to use the
+[OpenEuropa components](https://github.com/openeuropa/openeuropa/blob/master/docs/openeuropa-components.md),
+community contributed [Drupal modules](https://www.drupal.org/project/project_module)
+and [PHP packages](https://packagist.org/explore/).
```
Comment (ademarco) [inline README.md]: Our QA colleagues are actually working on restricting the number of modules and PHP projects that site developers can use when building a new site, it's maybe better to just keep here the part regarding the OpenEuropa components until further notice from the QA team.

File: README.md [hunk 14]
```diff
@@ -1,2 +1,126 @@
...
+The final step is to create a new git repository and commit all the files. A
+`.gitignore` file is provided to ensure you only commit your own project files.
+
+```
+$ git init
+$ git add .
+$ git commit -m "Initial commit."
+```
+
+Now you are ready to push your project to your chosen code hosting service.
```
Comment (ademarco) [inline README.md]: Shouldn't we add somewhere a part regarding how to work with site configuration? Like after installation site developers will probably want to run a `drush cex` and commit / push what will be dumped in `./config/sync`.

Comment (pfrenssen) [reply README.md]: There is a separate ticket for this: https://webgate.ec.europa.eu/CITnet/jira/browse/OPENEUROPA-799

File: scripts/composer/SetupWizard.php (hunk 10 above)
Comment (pfrenssen) [reply scripts/composer/SetupWizard.php]: There is a small problem remaining with this, I haven't found how to reload the {{composer.json}} after modifying it. At that point the unmodified version has already been statically cached in memory by Composer. Later on it uses the unmodified version to generate the autoloader, so we now have to manually execute `composer dump-autoload` after running the wizard.

File: README.md (hunk 14 above)
Comment (ademarco) [reply README.md]: Ah I see! I've linked OPENEUROPA-9 to that and added it to the correct epic.

File: composer.json (hunk 8 above)
Comment (pfrenssen) [reply composer.json]: This is not a component. This builds actual websites that are going in production at some point, so we need a place outside the docroot to store files that should not be accessible over the internet, like the `vendor/` folder, `config/sync` etc.

File: composer.json (hunk 6 above)
Comment (pfrenssen) [reply composer.json]: I don't think this is a good idea. Clearly none of our components are affected by this problem if they are currently working without this patch.

So they are not broken, why fix them?

File: scripts/composer/SetupWizard.php (hunk 11 above)
Comment (pfrenssen) [reply scripts/composer/SetupWizard.php]: It might be possible in the later `cleanup()` step maybe? Another possibility is to run `composer dump-autoload` in the task runner prior to installation.

File: composer.json (hunk 8 above)
Comment (ademarco) [reply composer.json]: Yes, things like `vendor` and `config/sync`  will still be outside the docroot, what I'm suggesting is just to remove `web` and have docroot under `./build`.

File: composer.json (hunk 6 above)
Comment (ademarco) [reply composer.json]: The patch makes detection of strict type compatible with php-parser ~4 but we use php-parser `~2.1|~3.0` here, so what's the use of having this patch applied here?

File: scripts/composer/SetupWizard.php (hunk 11 above)
Comment (ademarco) [reply scripts/composer/SetupWizard.php]: Yeah I'd try to put it in `cleanup()` then, so we have the whole thing being taken care in this composer plugin.

Comment (imanoleguskiza) [reply scripts/composer/SetupWizard.php]: I just dump-autoload as a command under the post-install scripts on composer.json and it seems to do the trick, custom namespaces are being loaded properly.

File: composer.json (hunk 8 above)
Comment (imanoleguskiza) [reply composer.json]: I think further discussion needs to be done around this topic, there seem to be different (valid) opinions on the matter. I suggest we take this to a follow up ticket.

File: runner.yml.dist (hunk 9 above)
Comment (ademarco) [reply runner.yml.dist]: After a discussion with @pfrenssen we will tackle this at the Task Runner level, I've opened a follow up issue here: https://webgate.ec.europa.eu/CITnet/jira/browse/OPENEUROPA-1139

File: composer.json (hunk 8 above)
Comment (ademarco) [reply composer.json]: After discussing it with @pfrenssen we will use `web` for the time being.

File: composer.json [hunk 15]
```diff
@@ -18,15 +18,16 @@
...
         "openeuropa/code-review": "~0.3",
         "openeuropa/composer-artifacts": "~0.1",
         "openeuropa/task-runner": "~0.5",
         "webflo/drupal-core-require-dev": "~8.6"
     },
     "scripts": {
         "post-install-cmd": [
-            "./vendor/bin/run drupal:site-setup",
-            "composer dump-autoload"
+            "@drupal-scaffold",
```
Comment (pfrenssen) [inline composer.json]: Scaffolding should not be executed on `composer install`. Since the scaffolded files are committed it is pointless to regenerate them on install. See the documentation for drupal-composer/drupal-scaffold: https://github.com/drupal-composer/drupal-scaffold/blob/master/README.md

> Once drupal-scaffold is required by your project, it will automatically update your scaffold files whenever composer update changes the version of drupal/core installed.

File: composer.json [hunk 16]
```diff
@@ -35,11 +36,12 @@
             "@cleanup"
         ],
         "post-update-cmd": [
-            "./vendor/bin/run drupal:site-setup",
-            "composer dump-autoload"
+            "@drupal-scaffold",
```
Comment (pfrenssen) [inline composer.json]: This line is only needed when using drupal-composer/drupal-scaffold versions 2.4 and lower. Since version 2.5 the project is doing this automatically. We have the right version requirement.

File: composer.json [hunk 17]
```diff
@@ -35,11 +36,12 @@
...
+            "./vendor/bin/run drupal:site-setup"
         ],
         "cleanup": "DrupalSiteTemplate\\composer\\SetupWizard::cleanup",
-        "setup": "DrupalSiteTemplate\\composer\\SetupWizard::setup"
+        "setup": "DrupalSiteTemplate\\composer\\SetupWizard::setup",
+        "drupal-scaffold": "DrupalComposer\\DrupalScaffold\\Plugin::scaffold"
```
Comment (pfrenssen) [inline composer.json]: This line is not needed since we are using drupal-composer/drupal-scaffold ^2.5.

File: composer.json [hunk 18]
```diff
@@ -18,15 +18,16 @@
     "require-dev": {
         "drupal/console": "^1.0.2",
         "drupal/drupal-extension": "~4.0",
+        "nikic/php-parser": "~3",
```
Comment (pfrenssen) [inline composer.json]: We don't directly depend on PHP-parser, why is this added here?

File: composer.json [hunk 19]
```diff
@@ -52,16 +54,11 @@
...
         "installer-paths": {
-            "build/web/core": ["type:drupal-core"],
-            "build/web/libraries/{$name}": ["type:drupal-library"],
-            "build/web/modules/contrib/{$name}": ["type:drupal-module"],
-            "build/web/profiles/contrib/{$name}": ["type:drupal-profile"],
-            "build/web/themes/contrib/{$name}": ["type:drupal-theme"]
-        },
-        "patches": {
-            "phpro/grumphp": {
-                "Make detection of strict types compatible with PhpParser 4 - https://github.com/phpro/grumphp/pull/525": "https://patch-diff.githubusercontent.com/raw/phpro/grumphp/pull/525.patch"
```
Comment (pfrenssen) [inline composer.json]: This hasn't been released yet, I would keep the patch for now. Once this is released we can update version requirements in the code review tool and then depend on the new version of code review.

File: .gitignore [hunk 20]
```diff
@@ -5,7 +5,10 @@
 /.idea/
 
 # Ignore the build.
-/build/
+/web/
```
Comment (pfrenssen) [inline .gitignore]: As discussed yesterday, we should no longer ignore the web root since this contains the scaffolded files and these should be committed. Let's also move the folders for the modules into the right location and no longer symlink it.

See the documentation for drupal-composer/drupal-scaffold: https://github.com/drupal-composer/drupal-scaffold/blob/master/README.md

> When using Composer to install or update the Drupal development branch, the scaffold files are always taken from the HEAD of the branch (or, more specifically, from the most recent development .tar.gz archive). This might not be what you want when using an old development version (e.g. when the version is fixed via composer.lock). **To avoid problems, always commit your scaffold files to the repository any time that composer.lock is committed.** Note that the correct scaffold files are retrieved when using a tagged release of drupal/core (recommended).

Let's add these folders to .gitignore (ref. https://github.com/drupal-composer/drupal-project/blob/8.x/.gitignore)

```
/web/core/
/web/modules/contrib/
/web/themes/contrib/
/web/profiles/contrib/
/web/libraries/

# Ignore sensitive information
/web/sites/*/settings.php
/web/sites/*/settings.local.php

# Ignore Drupal's file directory
/web/sites/*/files/

# Ignore SimpleTest multi-site environment.
/web/sites/simpletest
```

File: scripts/composer/SetupWizard.php [hunk 21]
```diff
@@ -15,6 +15,12 @@ class SetupWizard {
   /**
    * The setup wizard.
    *
+   * PHP Mess Detector is throwing unreasonable errors.
+   * Remove the following @SuppressWarnings once OPENEUROPA-792 is fixed.
+   * @link https://webgate.ec.europa.eu/CITnet/jira/browse/OPENEUROPA-792
+   *
+   * @SuppressWarnings(PHPMD.LongVariable)
```
Comment (pfrenssen) [inline scripts/composer/SetupWizard.php]: Good idea moving this here.

File: .gitignore (hunk 20 above)
Comment (pfrenssen) [reply .gitignore]: We should also document this clearly in the README so our users are aware that they should commit the scaffolding files. This can be done as part of OPENEUROPA-800.

File: composer.json (hunk 18 above)
Comment (pfrenssen) [reply composer.json]: It was added as a workaround to avoid the patch for GrumPHP.

It's better to have the patch I think. Having a patch clearly communicates to the end user "We are affected by this particular bug in this package, here is the link to the problem".

While having this version declared here communicates "We are depending on PHP Parser version 3 for development purposes". There is no way for an end user to know that this dependency on PHP Parser is working around a bug in GrumPHP.

There is another advantage to having the patch: once a new version of GrumPHP is released that contains the patch we will be alerted immediately since the patch will fail to apply. That will be the right moment to roll a new release for the code review tool and removing the patch here.

**Comment by ademarco** [comment]
@pfrenssen don't we need to explicitly require `nikic/php-parser` as a dev dependency? We do need that library to perform the following `grumphp.yml.dist` task:

```
  extra_tasks:
    phpparser:
      ignore_patterns: %tasks.phpcs.ignore_patterns%
      visitors:
        declare_strict_types: ~
      triggered_by: %tasks.phpcs.triggered_by%
```

And, at the moment, it we get it there since it's required by a Drush dependency, see below:

```
$ composer depends nikic/php-parser
psy/psysh  v0.9.7  requires  nikic/php-parser (~1.3|~2.0|~3.0|~4.0)  

$ composer depends psy/psysh  
drupal/console  1.8.0  requires  psy/psysh (0.6.* || ~0.8)  
drush/drush     9.3.0  requires  psy/psysh (~0.6)           
```

While we should explicitly require all our dependencies since we cannot be sure that our dependencies will not stop depending on `nikic/php-parser` one day.

Also, if we set major version to `~3` we don't need to add the patch.

**Comment by pfrenssen** [comment]
The patch has been removed again, put this back please. Also we are not limited to PHP Parser 3.x, we support `~2.0|~3.0|~4.0`.
